### PR TITLE
IPS-1094: Canary deployments

### PR DESCRIPTION
**Pull request by louiseli1**
### What changed

Files changed:

[infrastructure/private-api.yaml](https://github.com/govuk-one-login/ipv-cri-otg-hmrc/compare/IPS-1094/step-function-canaries?expand=1#diff-c98afc506962a43221fbfdc3d8330748f4ca867565692a1d497aa8cd39f72dc0)
- Edited api spec to point to live alias of state machine

[infrastructure/template.yaml](https://github.com/govuk-one-login/ipv-cri-otg-hmrc/compare/IPS-1094/step-function-canaries?expand=1#diff-1d6a0a19507d7cdddd8ff209f817d2d18f3fb22ceec6b0eb332b0a14292c69b0)
- Added `LambdaDeploymentPreference` and `StepFunctionsDeploymentPreference` parameters
- Add deployment preference to `BearerTokenRetrievalStateMachine` and `LogRedactionFunction`
- Removed references to `SecretsManagerHandlerFunction` from `BearerTokenRetrievalStateMachine`
- Added `CodeDeployServiceRole` for canaries

Changes to alarms:
- `BearerTokenRetrievalStateMachine`:
  - Updated MetricFilter (previously not in use) to search for `TaskFailed`
  - Added `ExecutionFailed` canary alarm
- `LogRedactionFunction`:
  - Created lambda error canary alarm
  - Removed `DeployAlarms` condition from `FatalErrorAlarm` so it can be used for canaries

### Why did it change

- Canary deployments across identity pod

### Issue tracking

- [IPS-1094](https://govukverify.atlassian.net/browse/IPS-1094)

##### Canary deployment in progress
LogRedactionFunctionCanary

##### Canary deployment complete
LogRedactionFunctionCanaryFinished

##### Canary deployment rollback
LogRedactionFunctionCanaryRollback

#### Canary deployment in progress
image

#### Test rollback with API call to state machine (Based on TaskFailed, or any invalid tokenType)
Screenshot 2024-09-27 at 16 01 32

#### Task failed metric filter and alarm
Screenshot 2024-09-27 at 16 10 25

#### Rollback of StepFunction to original version
Screenshot 2024-09-27 at 16 03 32
Screenshot 2024-09-27 at 16 06 18

#### Next canary deployment (successful API call and execution)
Screenshot 2024-09-27 at 16 15 49
image

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

[IPS-1094]: https://govukverify.atlassian.net/browse/IPS-1094?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: infrastructure/template.yaml [hunk 1]
```diff
@@ -1190,14 +1220,14 @@ Resources:
       LogGroupName: !Ref RedactedBearerTokenRetrievalStateMachineLogGroup
       DestinationArn: !FindInMap [PlatformConfiguration, !Ref Environment, CSLSEGRESS]
 
-  BearerTokenRetrievalStateMachineFailedMetric:
+  BearerTokenRetrievalStateMachineTaskFailedMetric:
     Type: AWS::Logs::MetricFilter
     Properties:
       LogGroupName: !Ref BearerTokenRetrievalStateMachineLogGroup
-      FilterPattern: '{($.type = "ExecutionFailed")}'
+      FilterPattern: '{$.type = "TaskFailed"}'
```
Comment (Caitlin-cooling) [inline infrastructure/template.yaml:1227]: How come we changed this? My understanding is that `TaskFailed` doesn't cover timeouts?

Comment (Caitlin-cooling) [reply infrastructure/template.yaml:1227]: Ignore me you've got `BearerTokenRetrievalStateMachineExecutionFailedCanaryAlarm

**Review by Caitlin-cooling — APPROVED** [review]

File: infrastructure/template.yaml [hunk 2]
```diff
@@ -1224,6 +1254,47 @@ Resources:
...
+        - !ImportValue platform-alarm-warning-alert-topic
+      AlarmName: !Sub ${AWS::StackName}-${Environment}-BearerTokenRetrieval-ExecutionFailedCanaryAlarm
+      AlarmDescription: !Sub "ExecutionFailed error returned from the BearerTokenRetrievalStateMachine"
+      ComparisonOperator: GreaterThanThreshold
+      EvaluationPeriods: 1
+      MetricName: ExecutionsFailed
+      Namespace: AWS/States
+      Period: 60
+      Statistic: Sum
+      Threshold: 0
```
Comment (sandragds) [inline infrastructure/template.yaml:1273]: should be 1?

**Comment by sandragds** [comment]
is it worth updating a readme to include canary work e.g. - https://github.com/govuk-one-login/ipv-cri-check-hmrc-api/pull/412/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R77-R90

File: infrastructure/template.yaml (hunk 2 above)
Comment (louiseli1) [reply infrastructure/template.yaml:1273]: `ComparisonOperator: GreaterThanThreshold` so 0 is technically fine here - however, I've updated just for consistency! (there's a mix of both in the template)

File: README.md [hunk 3]
```diff
@@ -1,3 +1,18 @@
...
+
+Canary deployments will cause a rollback if any canary alarms associated with a lambda or step-functions are triggered and a slack notification will be sent to #di-orange-warnings-non-prod or #di-orange-warning-alerts-prod.
+
+To skip canaries such as when releasing urgent changes to production, set the last commit message to contain either of these phrases: [skip canary], [canary skip], or [no canary] as specified in the [Canary Escape Hatch guide](https://govukverify.atlassian.net/wiki/spaces/PLAT/pages/3836051600/Rollback+Recovery+Guidance#Escape-Hatch%3A-how-to-skip-canary-deployments-when-needed).
+`git commit -m "some message [skip canary]"`
+
+Note: To update LambdaDeploymentPreference or StepFunctionsDeploymentPreference, update the LambdaCanaryDeployment or StepFunctionsDeploymentPreference pipeline parameter in the [identity-common-infra repository](https://github.com/govuk-one-login/identity-common-infra/tree/main/terraform/orange/hmrc-check). To update the LambdaDeploymentPreference or StepFunctionsDeploymentPreference for a stack in dev using sam deploy, parameter override needs to be set in the [deploy script](./deploy.sh):
```
Comment (sandragds) [inline README.md:15]: just the link needs to point to otg here so change to:
https://github.com/govuk-one-login/identity-common-infra/tree/main/terraform/orange/hmrc-otg

**Review by sandragds — APPROVED** [review]
Looks good!

**Review by sandragds — APPROVED** [review]
Looks good!

**Review by Caitlin-cooling — APPROVED** [review]

**Pull request merged**